sigstore/models.py

```diff
@@ -187,6 +186,77 @@ def _from_response(cls, dict_: dict[str, Any]) -> LogEntry:
             inclusion_promise=entry["verification"]["signedEntryTimestamp"],
         )
 
+    @classmethod
+    def _from_dict_rekor(cls, dict_: dict[str, Any]) -> LogEntry:
+        """
+        Create a new `LogEntry` from the given Rekor TransparencyLogEntry.
+        """
+        tlog_entry = rekor_v1.TransparencyLogEntry()
+        tlog_entry.from_dict(dict_)
+
+        inclusion_proof: InclusionProof | None = tlog_entry.inclusion_proof
+        # This check is required by us as the client, not the
+        # protobuf-specs themselves.
+        if inclusion_proof is None or inclusion_proof.checkpoint.envelope is None:
+            raise InvalidBundle("entry must contain inclusion proof")
+
+        parsed_inclusion_proof = LogInclusionProof(
+            checkpoint=inclusion_proof.checkpoint.envelope,
+            hashes=[h.hex() for h in inclusion_proof.hashes],
+            log_index=inclusion_proof.log_index,
+            root_hash=inclusion_proof.root_hash.hex(),
+            tree_size=inclusion_proof.tree_size,
+        )
+
+        return LogEntry(
+            uuid=None,
+            body=B64Str(base64.b64encode(tlog_entry.canonicalized_body).decode()),
+            integrated_time=tlog_entry.integrated_time,
+            log_id=tlog_entry.log_id.key_id.hex(),
+            log_index=tlog_entry.log_index,
+            inclusion_proof=parsed_inclusion_proof,
+            inclusion_promise=B64Str(
+                base64.b64encode(
+                    tlog_entry.inclusion_promise.signed_entry_timestamp
+                ).decode()
+            ),
+        )
+
+    def _to_dict_rekor(self, is_message_signature: bool) -> dict[str, Any]:
```

Comment:
Rather than passing in `is_message_signature`, I think we should infer this from `self.body` :slightly_smiling_face:

Reply:
Got it. I see that body has the following structure:
```json
{
    "apiVersion": "0.0.1",
    "kind": "hashedrekord",
    "spec": {"..."}
}
```

which corresponds to the [`ProposedEntry`](https://github.com/trailofbits/sigstore-rekor-types/blob/769065354dcd3f8fac175fda6c8dcd0df562dd7b/rekor_types/__init__.py#L127-L130) union type in `sigstore-rekor-types`.

These can be one of several [types](https://github.com/sigstore/rekor/tree/main/pkg/types). Our previous logic said that :
```python
        if isinstance(content, common_v1.MessageSignature):
            inner.message_signature = content
            tlog_entry.kind_version = rekor_v1.KindVersion(
                kind="hashedrekord", version="0.0.1"
            )
        else:
            inner.dsse_envelope = content._inner
            tlog_entry.kind_version = rekor_v1.KindVersion(kind="dsse", version="0.0.1")
```

How would that correspond to the different types that a `ProposedEntry` can be? What I'm asking is which `kind` in the body (`hashedrekord`, `cose`, `alpine`, `dsse`, etc) corresponds to the two kinds we were handling above (`hashedrekord` and `dsse`)

Reply:
The two are mirrored, but sigstore-python (intentionally) only supports `hashedrekord` and `dsse` for now. So I think what we want here is to fail if we see any type other than those two (which should be impossible, since we should also reject them earlier in the stack).

Reply:
done!